decode.go

```diff
@@ -0,0 +1,292 @@
+			return readString(r, val)
+
+		case reflect.Ptr:
+			return d.decode(r, val.Elem())
+
+		case reflect.Struct:
+			return d.readStruct(r, val)
+
+		case reflect.Slice:
+			return d.readSlice(r, val)
+
+		default:
+			return fmt.Errorf("unsupported type %s", val.Type())
+		}
+	}
+}
+
+func (d *Decoder) readStruct(r io.Reader, val reflect.Value) error {
+	for i := 0; i < val.NumField(); i++ {
+		f := val.Field(i)
+
+		// if the field is a pointer we need to create
+		// the value before we can marshal data into it.
+		if f.Type().Kind() == reflect.Ptr {
+			f.Set(reflect.New(f.Type().Elem()))
+		}
+
+		if err := d.decode(r, f); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *Decoder) readSlice(r io.Reader, val reflect.Value) error {
+	v, err := readUint(r, 4)
+	if err != nil {
+		return err
+	}
+
+	n := uint32(v)
+	if n == null {
+		return nil
+	}
+
+	if n > math.MaxInt32 {
+		return fmt.Errorf("array too large: %d", n)
+	}
+
+	// elemType is the type of the slice elements
+	// e.g. *Foo for []*Foo
+	elemType := val.Type().Elem()
+	// fmt.Println("elemType: ", elemType.String())
+
+	// fast path for []byte
+	if elemType.Kind() == reflect.Uint8 {
+		// fmt.Println("decode: []byte fast path")
+		b := make([]byte, int(n))
+		m, err := r.Read(b)
+		if err != nil {
+			return err
+		}
+		if m < int(n) {
+			return io.ErrUnexpectedEOF
+		}
+		val.SetBytes(b)
+		return nil
+	}
+
+	// a is a slice of []*Foo
+	a := reflect.MakeSlice(val.Type(), int(n), int(n))
+	for i := 0; i < int(n); i++ {
+
+		// if the slice elements are pointers we need to create
+		// them before we can marshal data into them.
+		if elemType.Kind() == reflect.Ptr {
+			a.Index(i).Set(reflect.New(elemType.Elem()))
+		}
+
+		if err := d.decode(r, a.Index(i)); err != nil {
+			return err
+		}
+	}
+	val.Set(a)
+
+	return nil
+}
+
+func readInt(r io.Reader, n int) (int64, error) {
+	v, err := readUint(r, n)
+	if err != nil {
+		return 0, err
+	}
+	switch n {
+	case 1:
+		return int64(int8(v)), nil
+	case 2:
+		return int64(int16(v)), nil
+	case 4:
+		return int64(int32(v)), nil
+	case 8:
+		return int64(v), nil
+	default:
+		return 0, fmt.Errorf("invalid int width: %d", n)
+	}
+}
+
+func readUint(r io.Reader, n int) (uint64, error) {
+	b, err := read(r, n)
+	if err != nil {
+		return 0, err
+	}
+	switch n {
+	case 1:
+		return uint64(b[0]), nil
+	case 2:
+		return uint64(binary.LittleEndian.Uint16(b)), nil
+	case 4:
+		return uint64(binary.LittleEndian.Uint32(b)), nil
+	case 8:
+		return binary.LittleEndian.Uint64(b), nil
+	default:
+		return 0, fmt.Errorf("invalid uint width: %d", n)
+	}
+}
+
+func readString(r io.Reader, val reflect.Value) error {
+	v, err := readUint(r, 4)
+	if err != nil {
+		return err
+	}
+	n := uint32(v)
+	if n == null {
+		return nil
+	}
+	b, err := read(r, int(n))
+	if err != nil {
+		return err
+	}
+	val.SetString(string(b))
+	return nil
+}
+
+func read(r io.Reader, n int) ([]byte, error) {
+	b := make([]byte, n)
+	p := 0
+	for p < n {
```

Comment:
this can be replaced with `io.ReadFull`

Reply:
Yes :) Should have had another look at `io`.